codegen/src/types/composite_def.rs

```diff
@@ -0,0 +1,320 @@
+    pub fields: CompositeDefFields,
+}
+
+impl CompositeDef {
+    /// Construct a definition which will generate code for a standalone `struct`.
+    pub fn struct_def(
+        ident: &str,
+        type_params: TypeDefParameters,
+        fields_def: CompositeDefFields,
+        field_visibility: Option<syn::Visibility>,
+        type_gen: &TypeGenerator,
+    ) -> Self {
+        let mut derives = type_gen.derives().clone();
+        let fields = fields_def.fields();
+
+        if fields.len() == 1 {
+            // any single field wrapper struct with a concrete unsigned int type can derive
+            // CompactAs.
+            let field = &fields[0];
+            if !type_params
+                .params()
+                .iter()
+                .any(|tp| Some(tp.original_name.to_string()) == field.type_name)
+            {
+                let ty = type_gen.resolve_type(field.type_id);
+                if matches!(
+                    ty.type_def(),
+                    TypeDef::Primitive(
+                        TypeDefPrimitive::U8
+                            | TypeDefPrimitive::U16
+                            | TypeDefPrimitive::U32
+                            | TypeDefPrimitive::U64
+                            | TypeDefPrimitive::U128
+                    )
+                ) {
+                    derives.push_codec_compact_as()
+                }
+            }
+        }
+
+        let name = format_ident!("{}", ident.to_camel_case());
+
+        Self {
+            name,
+            kind: CompositeDefKind::Struct {
+                derives,
+                type_params,
+                field_visibility,
+            },
+            fields: fields_def,
+        }
+    }
+
+    /// Construct a definition which will generate code for an `enum` variant.
+    pub fn enum_variant_def(ident: &str, fields: CompositeDefFields) -> Self {
+        let name = format_ident!("{}", ident);
+        Self {
+            name,
+            kind: CompositeDefKind::EnumVariant,
+            fields,
+        }
+    }
+}
+
+impl quote::ToTokens for CompositeDef {
+    fn to_tokens(&self, tokens: &mut TokenStream) {
+        let name = &self.name;
+
+        let decl = match &self.kind {
+            CompositeDefKind::Struct {
+                derives,
+                type_params,
+                field_visibility,
+            } => {
+                let unused_phantom_marker = type_params.unused_params_phantom_data();
+                let fields = self.fields.field_tokens(
+                    field_visibility.as_ref(),
+                    unused_phantom_marker,
+                    true,
+                );
+
+                quote! {
+                    #derives
+                    pub struct #name #type_params #fields
+                }
+            }
+            CompositeDefKind::EnumVariant => {
+                let fields = self.fields.field_tokens(None, None, false);
```

Comment:
I wonder if it would make sense to have separate `field_tokens` methods for Struct vs Enums, e.g. `struct_tokens()`/`enum_tokens()`, to make the signature a bit easier on the eyes here.

Reply:
Done, let me know if that works now.